Adding reactivesocket-mime-types module.

NiteshKant opened the pull request:
This module provides support for encoding/decoding ReactiveSocket data and metadata into using different mime types as defined by [ReactiveSocket protocol](https://github.com/ReactiveSocket/reactivesocket/blob/master/Protocol.md#setup-frame).
The support for mime types is not comprehensive but it will atleast support the [default metadata mime type](https://github.com/ReactiveSocket/reactivesocket/blob/mimetypes/MimeTypes.md)

See README.md for usage.

NiteshKant commented:
@rdegnan @stevegury @robertroeser PTAL

stevegury commented on reactivesocket-mime-types/README.md:
```diff
@@ -0,0 +1,66 @@
+## Overview
+
+This module provides support for encoding/decoding ReactiveSocket data and metadata into using different mime types as defined by [ReactiveSocket protocol](https://github.com/ReactiveSocket/reactivesocket/blob/master/Protocol.md#setup-frame).
+The support for mime types is not comprehensive but it will atleast support the [default metadata mime type](https://github.com/ReactiveSocket/reactivesocket/blob/mimetypes/MimeTypes.md)
```
at least

NiteshKant replied:
_Grammer_ Nazis ;)
Will change

stevegury commented on reactivesocket-mime-types/src/main/java/io/reactivesocket/mimetypes/KVMetadata.java:
```diff
@@ -0,0 +1,58 @@
+     *
+     * @return A read-only set of all entries in this metadata.
+     */
+    Set<Entry<String, ByteBuffer>> entries();
```
Why the `Set`, it should be a list, shouldn't it?

NiteshKant replied:
You mean, we need to support duplicate keys?

stevegury replied:
If you don't to support duplicate keys, then a `Map` is a better interface.

yschimke replied:
Also Entry equality is on key and value, so this is currently a MultiMap

NiteshKant replied:
@stevegury I do not intend to support duplicate keys. Also, CBOR does not recommend duplicate keys and may cause decoding failure for some strict parsers. Ref [here](https://tools.ietf.org/html/rfc7049#section-3.7)
I guess we can use `Map` interface. I do not have strong preference for either, this is just a small API surface. This method is actually just as `Map.entrySet()`

@yschimke Duplicate keys vs `MultiMap`, same difference?

stevegury commented on reactivesocket-mime-types/src/main/java/io/reactivesocket/mimetypes/internal/ByteBufferInputStream.java:
```diff
@@ -0,0 +1,59 @@
+ * This code is copied from {@link com.fasterxml.jackson.databind.util.ByteBufferBackedInputStream} with the
+ * modifications to make the stream mutable per thread.
+ */
+public final class ByteBufferInputStream extends InputStream {
```
(Here and elsewhere)
Maybe it's worth adding a `clear()` method to get rid of the ThreadLocal ref to the buffer.
Buffers can be large, and some application uses lots of threads...
What do you think?

NiteshKant replied:
This pattern (thread local caches) really blows up when it is used in places outside eventloops. So, the encoding/decoding should always happen on the eventloop. 
(This is the reason, RxNetty goes out of its way to make sure that there isn't an easy way to allocate buffers out of the eventloops)

I am fine with adding a `clear()` method but this is an internal class and used by codecs, so if we add a lifecycle method, we will have to tie it to some lifecycle somewhere. If the lifecycle is for the thread then I guess it will be GC'ed, rite?

stevegury commented:
Small nit, but I think a better name for the module would be `reactivesocket-default-encoding`.
Indeed, the module contains more that just MIME types.
Not sure if some of those interfaces could be moved to `reactivesocket-java`.

NiteshKant commented:
@stevegury I started off with that module name but since it has general codec feature, i changed it. 
It could be reactivesocket-codecs but then it can be confused with framing codecs 😵 

We can probably move `MimeType` interface to reactivesocket-java.

NiteshKant commented:
I have incorporated the online and offline review comments. Following are the changes:

- `KVMetadata` now extends `Map`.
- `ByteBufferInput/OutputStream` is now package local. The `ThreadLocal` references are now moved out of the class to `AbstractJacksonCodec`.

stevegury commented:
👍
